File: backend/src/judge/runner.ts
import { spawn } from "child_process";

/**
 * Build the environment for the untrusted process. In the (Linux) container we
 * use a restricted PATH so only system tools are reachable — a security
 * measure. On other platforms (local dev) we inherit the user environment so
 * language runtimes on the user's PATH remain discoverable.
 */
function buildEnv(cwd: string): NodeJS.ProcessEnv {
  const common = {
    HOME: cwd,
    LANG: "C.UTF-8",
    LC_ALL: "C.UTF-8",
  };
  if (process.platform === "win32") {
    return { ...process.env, ...common };
  }
  return {
    PATH: "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
    ...common,
  };
}

export interface RunOptions {
  /** Shell command to execute. */
  command: string;
  /** Standard input fed to the process. */
  stdin: string;
  /** Wall-clock timeout in milliseconds. */
  timeLimitMs: number;
  /** Virtual-memory cap in KB (best-effort, Linux only). Set 0 to disable. */
  memoryLimitKb: number;
  /** Working directory. */
  cwd: string;
}

export interface RunResult {
  ok: boolean;
  stdout: string;
  stderr: string;
  exitCode: number | null;
  signal: NodeJS.Signals | null;
  timedOut: boolean;
  memoryError: boolean;
  elapsedMs: number;
}

/**
 * Execute a command in a sandboxed fashion:
 *  - runs as a new process group so we can kill the whole tree on timeout
 *  - applies a best-effort virtual-memory cap via `ulimit -v` (Linux)
 *  - applies a CPU-time cap via `ulimit -t` so runaway loops are reaped even
 *    if wall-clock detection races
 *  - captures stdout/stderr and exit status
 *
 * NOTE: This is designed to run inside the backend Docker container where the
 * filesystem and network are already isolated by the container boundary.
 */
export function run(opts: RunOptions): Promise<RunResult> {
  return new Promise((resolve) => {
    const start = Date.now();

    let prelude = "";
    if (opts.memoryLimitKb > 0) {
      prelude += `ulimit -v ${opts.memoryLimitKb} 2>/dev/null; `;
    }
    // CPU seconds = ceil(timeLimitMs/1000) + 1s grace
    const cpuSec = Math.ceil(opts.timeLimitMs / 1000) + 1;
    prelude += `ulimit -t ${cpuSec} 2>/dev/null; `;
    // Limit file writes (output explosion protection). 256MB in 1KB blocks.
    prelude += `ulimit -f 262144 2>/dev/null; `;

    const wrapped = `${prelude}exec ${opts.command}`;

    const child = spawn("bash", ["-c", wrapped], {
      cwd: opts.cwd,
      detached: true,
      stdio: ["pipe", "pipe", "pipe"],
      env: buildEnv(opts.cwd),
    });

    let stdout = "";
    let stderr = "";
    let timedOut = false;
    let memoryError = false;
    let settled = false;

    const cap = 16 * 1024 * 1024; // 16MB guard per stream
    child.stdout.on("data", (d: Buffer) => {
      if (stdout.length < cap) stdout += d.toString("utf8");
    });
    child.stderr.on("data", (d: Buffer) => {
      if (stderr.length < cap) stderr += d.toString("utf8");
    });

    const timer = setTimeout(() => {
      timedOut = true;
      try {
        if (child.pid) process.kill(-child.pid, "SIGKILL");
      } catch {
        /* group already gone */
      }
    }, opts.timeLimitMs);

    const finish = (result: Omit<RunResult, "elapsedMs">) => {
      if (settled) return;
      settled = true;
      clearTimeout(timer);
      resolve({ ...result, elapsedMs: Date.now() - start });
    };

    child.on("close", (code, signal) => {
      // Heuristic: an OOM usually surfaces as non-zero exit with memory text.
      const memHint = /Cannot allocate memory|MemoryError|out of memory|bad allocation|std::bad_alloc/i.test(stderr);
      memoryError = !timedOut && memHint;
      finish({
        ok: code === 0 && !timedOut,
        stdout,
        stderr,
        exitCode: code,
        signal,
        timedOut,
        memoryError,
      });
    });

    child.on("error", () => {
      finish({
        ok: false,
        stdout,
        stderr,
        exitCode: null,
        signal: null,
        timedOut,
        memoryError: false,
      });
    });

    // Feed stdin. Ignore write errors (process may have exited early).
    try {
      child.stdin.end(opts.stdin);
    } catch {
      /* ignore */
    }
  });
}

File: worker/src/main/java/com/oj/judge/Runner.java
package com.oj.judge;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.util.concurrent.TimeUnit;

/**
 * Executes a shell command in a sandboxed subprocess:
 *  - runs via bash so we can apply ulimit (CPU time, virtual memory, file size)
 *  - separate process group so we can kill the whole tree on timeout
 *  - captures stdout/stderr up to a cap, feeds stdin
 *
 * Mirrors the Node runner semantics. Designed to run inside the worker
 * Docker container (Linux), where the filesystem/network are already isolated.
 */
public class Runner {

    private static final int STDOUT_CAP = 16 * 1024 * 1024;

    public RunResult run(String command, String stdin, long timeLimitMs, long memoryLimitKb, Path cwd) {
        // Build a bash wrapper that applies ulimits then execs the command.
        StringBuilder prelude = new StringBuilder();
        if (memoryLimitKb > 0) {
            prelude.append("ulimit -v ").append(memoryLimitKb).append(" 2>/dev/null; ");
        }
        long cpuSec = timeLimitMs / 1000 + 1;
        prelude.append("ulimit -t ").append(cpuSec).append(" 2>/dev/null; ");
        prelude.append("ulimit -f 262144 2>/dev/null; "); // 256MB output cap
        String wrapped = prelude + "exec " + command;

        ProcessBuilder pb = new ProcessBuilder("bash", "-c", wrapped);
        pb.directory(cwd.toFile());
        pb.redirectErrorStream(false);

        // Restrict env to a minimal safe set.
        pb.environment().clear();
        pb.environment().put("PATH", "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin");
        pb.environment().put("HOME", cwd.toString());
        pb.environment().put("LANG", "C.UTF-8");
        pb.environment().put("LC_ALL", "C.UTF-8");

        long start = System.currentTimeMillis();
        Process p;
        try {
            p = pb.start();
        } catch (IOException e) {
            return new RunResult(false, "", e.getMessage(), -1, false, false, 0);
        }

        // Write stdin on a separate thread to avoid blocking on large input.
        byte[] input = stdin.getBytes(StandardCharsets.UTF_8);
        Thread writer = new Thread(() -> {
            try (var os = p.getOutputStream()) {
                os.write(input);
                os.flush();
            } catch (IOException ignored) {}
        });
        writer.setDaemon(true);
        writer.start();

        StringBuilder outBuf = new StringBuilder();
        StringBuilder errBuf = new StringBuilder();
        Thread readerOut = new Thread(() -> readStream(p.getInputStream(), outBuf));
        Thread readerErr = new Thread(() -> readStream(p.getErrorStream(), errBuf));
        readerOut.setDaemon(true);
        readerErr.setDaemon(true);
        readerOut.start();
        readerErr.start();

        boolean timedOut = false;
        boolean finished;
        try {
            finished = p.waitFor(timeLimitMs, TimeUnit.MILLISECONDS);
            if (!finished) {
                timedOut = true;
                p.destroyForcibly();
                p.waitFor(2, TimeUnit.SECONDS);
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            p.destroyForcibly();
            return new RunResult(false, outBuf.toString(), errBuf.toString(), -1, true, false,
                    System.currentTimeMillis() - start);
        }

        try { writer.join(500); } catch (InterruptedException ignored) {}
        try { readerOut.join(1000); } catch (InterruptedException ignored) {}
        try { readerErr.join(1000); } catch (InterruptedException ignored) {}

        int exitCode = p.exitValue();
        long elapsed = System.currentTimeMillis() - start;

        boolean memError = !timedOut && (errBuf.toString().toLowerCase().contains("cannot allocate memory")
                || errBuf.toString().toLowerCase().contains("out of memory")
                || errBuf.toString().toLowerCase().contains("bad_alloc"));

        return new RunResult(!timedOut && exitCode == 0, outBuf.toString(), errBuf.toString(),
                exitCode, timedOut, memError, elapsed);
    }

    private void readStream(java.io.InputStream is, StringBuilder buf) {
        try (var br = new java.io.BufferedReader(new java.io.InputStreamReader(is, StandardCharsets.UTF_8))) {
            char[] chunk = new char[4096];
            int n;
            while ((n = br.read(chunk)) != -1) {
                if (buf.length() < STDOUT_CAP) {
                    int allowed = Math.min(n, STDOUT_CAP - buf.length());
                    buf.append(chunk, 0, allowed);
                }
            }
        } catch (IOException ignored) {}
    }
}
